contracts/compliance/modular/modules/MaxBalanceModule.sol

```diff
@@ -0,0 +1,204 @@
+
+    /**
+     *  this event is emitted when the max balance has been set for a compliance bound.
+     *  `_compliance` is the address of modular compliance concerned
+     *  `_maxBalance` is the max amount of tokens that a user can hold .
+     */
+    event MaxBalanceSet(address indexed _compliance, uint256 indexed _maxBalance);
+
+    event IDBalancePreSet(address indexed _compliance, address indexed _id, uint256 _balance);
+
+    /// errors
+    error MaxBalanceExceeded(address _compliance, uint256 _value);
+
+    /// functions
+
+    /**
+     *  @dev sets max balance limit for a bound compliance contract
+     *  @param _max max amount of tokens owned by an individual
+     *  Only the owner of the Compliance smart contract can call this function
+     *  emits an `MaxBalanceSet` event
+     */
+    function setMaxBalance(uint256 _max) external onlyComplianceCall {
+        _maxBalance[msg.sender] = _max;
+        emit MaxBalanceSet(msg.sender, _max);
+    }
+
+    /**
+     *  @dev make a batch transaction calling preSetModuleState multiple times
+     *  @param _compliance the address of the compliance contract to preset
+     *  @param _id the ONCHAINID address of the token holder
+     *  @param _balance the current balance of the token holder
+     *  Only the owner of the Compliance smart contract can call this function
+     *  emits _id.length `IDBalancePreSet` events
+     */
+    function batchPreSetModuleState(
+        address _compliance,
+        address[] calldata _id,
+        uint256[] calldata _balance) external {
+        for (uint i = 0; i < _id.length; i++) {
+            preSetModuleState(_compliance, _id[i], _balance[i]);
+        }
+    }
+
+    /**
+     *  @dev See {IModule-moduleTransferAction}.
+     *  no transfer action required in this module
+     */
+    function moduleTransferAction(address _from, address _to, uint256 _value) external override onlyComplianceCall {
+        address _idFrom = _getIdentity(msg.sender, _from);
+        address _idTo = _getIdentity(msg.sender, _to);
+        _IDBalance[msg.sender][_idTo] += _value;
+        _IDBalance[msg.sender][_idFrom] -= _value;
+        if (_IDBalance[msg.sender][_idTo] > _maxBalance[msg.sender]) revert MaxBalanceExceeded(msg.sender, _value);
+    }
+
+    /**
+     *  @dev See {IModule-moduleMintAction}.
+     *  no mint action required in this module
+     */
+    function moduleMintAction(address _to, uint256 _value) external override onlyComplianceCall {
+        address _idTo = _getIdentity(msg.sender, _to);
+        _IDBalance[msg.sender][_idTo] += _value;
+        if (_IDBalance[msg.sender][_idTo] > _maxBalance[msg.sender]) revert MaxBalanceExceeded(msg.sender, _value);
+    }
+
+    /**
+     *  @dev See {IModule-moduleBurnAction}.
+     *  no burn action required in this module
+     */
+    function moduleBurnAction(address _from, uint256 _value) external override onlyComplianceCall {
+        address _idFrom = _getIdentity(msg.sender, _from);
```

Comment:
You might want to check if the `_idFrom` is the null address (I think in our case here `_getIdentity()` returns the zero address if the identity was not found).

Reply:
done